crates/circuit_encodings/src/lib.rs

```diff
@@ -184,34 +175,66 @@ impl<
         let _ = self.push_and_output_intermediate_data(element, round_function);
     }
 
-    pub fn push_and_output_intermediate_data<
+    fn absorb_element<
         R: CircuitRoundFunction<F, AW, SW, CW> + AlgebraicRoundFunction<F, AW, SW, CW>,
         const AW: usize,
         const SW: usize,
         const CW: usize,
     >(
         &mut self,
-        element: I,
+        element_encoding: [F; N],
+        prev_commitment: [F; T],
         _round_function: &R,
     ) -> (
-        [F; T],                                    // old tail
-        QueueIntermediateStates<F, T, SW, ROUNDS>, // new head/tail, as well as round function ins/outs
+        [F; T],            // new commitment
+        [[F; SW]; ROUNDS], // intermediate states
     ) {
-        let old_tail = self.tail;
-        let encoding = element.encoding_witness();
         let mut to_hash = Vec::with_capacity(N + T);
-        to_hash.extend_from_slice(&encoding);
-        to_hash.extend(self.tail);
+        to_hash.extend_from_slice(&element_encoding);
+        to_hash.extend(prev_commitment);
 
         let mut state = R::initial_state();
         let states = absorb_multiple_rounds::<F, R, AbsorptionModeOverwrite, AW, SW, CW, ROUNDS>(
             &mut state, &to_hash,
         );
-        let new_tail =
+
+        let commitment =
             <R as AlgebraicRoundFunction<F, AW, SW, CW>>::state_into_commitment::<T>(&state);
+
+        (commitment, states)
+    }
+
+    pub fn make_round_function_pairs<
+        R: CircuitRoundFunction<F, AW, SW, CW> + AlgebraicRoundFunction<F, AW, SW, CW>,
+        const AW: usize,
+        const SW: usize,
+        const CW: usize,
+    >(
+        round_states: [[F; SW]; ROUNDS],
+        _round_function: &R,
+    ) -> [([F; SW], [F; SW]); ROUNDS] {
+        make_round_function_pairs(R::initial_state(), round_states)
+    }
+
+    pub fn push_and_output_intermediate_data<
+        R: CircuitRoundFunction<F, AW, SW, CW> + AlgebraicRoundFunction<F, AW, SW, CW>,
+        const AW: usize,
+        const SW: usize,
+        const CW: usize,
+    >(
+        &mut self,
+        element: I,
+        round_function: &R,
+    ) -> (
+        [[F; SW]; ROUNDS],                 // intermediate round states
```

Comment:
Are those still used anywhere? I remember comparing them for debug purposes on storage queue, but not much else

Reply:
It is used for debug here (storage queue, indeed):
https://github.com/matter-labs/zksync-protocol/pull/18/files/1c7423076fea8cb1e9d26cdbb012f2bf01868e9b#diff-8af2d01ce30e2a7584c906f9f1b7642e721043db0334bab6293dca4efe9e0a4eR273